beetsplug/lastgenre/__init__.py

```diff
@@ -123,27 +130,102 @@ def setup(self):
                 for line in f:
                     line = line.decode("utf-8").strip().lower()
                     if line and not line.startswith("#"):
-                        self.whitelist.add(line)
+                        whitelist.add(line)
+        return whitelist
 
-        # Read the genres tree for canonicalization if enabled.
-        self.c14n_branches = []
+    def _load_c14n_tree(self):
+        c14n_branches = []
         c14n_filename = self.config["canonical"].get()
-        self.canonicalize = c14n_filename is not False
-
+        canonicalize = c14n_filename is not False
         # Default tree
         if c14n_filename in (True, ""):
             c14n_filename = C14N_TREE
-        elif not self.canonicalize and self.config["prefer_specific"].get():
+        elif not canonicalize and self.config["prefer_specific"].get():
             # prefer_specific requires a tree, load default tree
             c14n_filename = C14N_TREE
-
         # Read the tree
         if c14n_filename:
             self._log.debug("Loading canonicalization tree {0}", c14n_filename)
             c14n_filename = normpath(c14n_filename)
             with codecs.open(c14n_filename, "r", encoding="utf-8") as f:
                 genres_tree = yaml.safe_load(f)
-            flatten_tree(genres_tree, [], self.c14n_branches)
+            flatten_tree(genres_tree, [], c14n_branches)
+        return c14n_branches, canonicalize
+
+    def _load_blacklist(self):
+        """Load the blacklist from a configured file path.
+
+        For maximum compatibility with regex patterns, a custom format is used:
+        - Each section starts with an artist name, followed by a colon.
+        - Subsequent lines are indented (at least one space, typically 4 spaces) and
+          contain a regex pattern to match a genre.
+
+
+        Supports a special '*' key in the blacklist for
+        global forbidden genres.
+
+        Example blacklist file format:
+            Artist Name:
+                .*rock.*
+                .*metal.*
+            Another Artist Name:
+                ^jazz$
+            *:
+                spoken word
+                comedy
+
+        Raises:
+            UserError: if the file format is invalid.
+        """
+        blacklist = defaultdict(list)
+        if not (bl_filename := self.config["blacklist"].get()):
+            return blacklist
+        if not os.path.isfile(bl_filename := normpath(bl_filename)):
+            self._log.error("Blacklist file not found: {} .", bl_filename)
+            return blacklist
+
+        self._log.debug("Loading blacklist file {0}", bl_filename)
+        section = None
+        with open(bl_filename, "r", encoding="utf-8") as f:
+            for lineno, line in enumerate(f, 1):
+                line = line.lower()
```

Comment:
this is save. everything should be handled lower case initially. later on literal strings get lower cased anyway (eg. is_forbidden method) and the regex patters use re.IGNORECAE anyway too.